crates/solver/src/settlement_submission.rs

```diff
@@ -9,23 +9,94 @@ use anyhow::{anyhow, Result};
 use contracts::GPv2Settlement;
 use ethcontract::{
     errors::{ExecutionError, MethodError},
-    Account, TransactionHash,
+    Account, Address, TransactionHash,
 };
 use futures::FutureExt;
-use gas_estimation::GasPriceEstimating;
+use gas_estimation::{EstimatedGasPrice, GasPriceEstimating};
 use primitive_types::{H256, U256};
 use shared::Web3;
 use std::{
-    sync::Arc,
+    collections::HashMap,
+    sync::{Arc, Mutex},
     time::{Duration, Instant},
 };
 use submitter::{
-    DisabledReason, Submitter, SubmitterGasPriceEstimator, SubmitterParams, TransactionSubmitting,
+    DisabledReason, Strategy, Submitter, SubmitterGasPriceEstimator, SubmitterParams,
+    TransactionHandle, TransactionSubmitting,
 };
 use web3::types::TransactionReceipt;
 
 const ESTIMATE_GAS_LIMIT_FACTOR: f64 = 1.2;
 
+// Key (Address, U256) represents pair (sender, nonce)
+type SubTxPool = HashMap<(Address, U256), Vec<(TransactionHandle, EstimatedGasPrice)>>;
+
+#[derive(Default)]
+struct TxPool {
+    eden: SubTxPool,
+    flashbots: SubTxPool,
+    custom_nodes: SubTxPool,
+}
```

Comment:
One thing I'm not too fond of in this design is that we can no longer have multiple strategies of the same time. For example, it could be possible to configure `StrategyArg::FlashbotsFast` to just be another instance of a `Flashbots` strategy with a different URL configured.

Maybe the submitter keeps track of its `TxPool` internally?

Reply:
Same as here: https://github.com/cowprotocol/services/pull/235#discussion_r886621071